src/test/ide.test.ts

```diff
@@ -804,4 +805,410 @@ test('we can find references to a dependency', async ({rig}) => {
+      },
+      {
+        label: 'deps',
+        kind: completionItemKinds.dependenciesOnly,
+      },
+      {
+        label: 'files',
+        kind: completionItemKinds.filesOnly,
+      },
+      {
+        label: 'foo',
+        kind: completionItemKinds.normalScript,
+      },
+      {
+        label: 'service',
+        kind: completionItemKinds.service,
+      },
+    ],
+  };
+  await assertCompletions(ide, {
+    path: rig.resolve('package.json'),
+    contentsWithPipe: JSON.stringify(
+      {
+        scripts: {
+          a: 'wireit',
+          b: 'wireit',
+        },
+        wireit: {
+          foo: {
+            command: 'echo',
+          },
+          bar: {
+            dependencies: ['fo|'],
+            command: 'echo',
+          },
+          service: {
+            service: true,
+            command: 'foo',
+          },
+          files: {
+            files: ['*.js'],
+          },
+          deps: {
+            dependencies: ['foo'],
+          },
+        },
+      },
+      null,
+      2,
+    ),
+    expected: expected,
+  });
+
+  await assertCompletions(ide, {
+    path: rig.resolve('package.json'),
+    contentsWithPipe: JSON.stringify(
+      {
+        scripts: {
+          a: 'wireit',
+          b: 'wireit',
+        },
+        wireit: {
+          foo: {
+            command: 'echo',
+          },
+          bar: {
+            dependencies: ['|'],
+            script: 'echo',
+          },
+          service: {
+            service: true,
+            command: 'foo',
+          },
+          files: {
+            files: ['*.js'],
+          },
+          deps: {
+            dependencies: ['foo'],
+          },
+        },
+      },
+      null,
+      2,
+    ),
+    expected: {
+      ...expected,
+      // It's incomplete because we don't know whether the user wants to type
+      // a ./ or a ../ or just the name of a script.
+      isIncomplete: true,
+    },
+  });
+});
+
+test('we can get completions for cross file dependencies', async ({rig}) => {
+  const ide = new IdeAnalyzer();
+  await rig.write('child/package.json', {
+    scripts: {
+      a: 'wireit',
+      b: 'wireit',
+    },
+    wireit: {
+      foo: {
+        command: 'echo',
+      },
+      bar: {
+        dependencies: ['foo'],
+      },
+      service: {
+        service: true,
+        command: 'foo',
+      },
+      files: {
+        files: ['*.js'],
+      },
+    },
+  });
+  const expected = {
+    // We actually propose all scripts, and let the IDE narrow them down.
+    isIncomplete: false,
+    items: [
+      {
+        label: 'a',
+        kind: completionItemKinds.normalScript,
+      },
+      {
+        label: 'b',
+        kind: completionItemKinds.normalScript,
+      },
+      {
+        label: 'bar',
+        kind: completionItemKinds.dependenciesOnly,
+      },
+      {
+        label: 'files',
+        kind: completionItemKinds.filesOnly,
+      },
+      {
+        label: 'foo',
```

Comment:
Given the suggestion `"foo"`, how does VSCode know to complete `"./child:fo|"` to `"./child:foo"` rather than to `"foo"` or `"./child:fofoo"`? I.e. how does it know to replace starting from after the `:`?

Edit: Ah, I see the answer, it must be the "triggerCharacters" setting. So it will go backwards from the cursor until it hits one of those, and then replace from that point?

Reply:
Yeah, but it's not great at it, particularly for scripts that have a colon in their name. We have all the info to do the right thing ourselves, but I struggle to express that using the LSP. See https://github.com/microsoft/vscode/issues/194580

Something to work on down the line